stl/inc/xstring

```diff
@@ -2722,35 +2722,62 @@ private:
         }
 
         _Tidy_deallocate_guard<basic_string> _Guard{this};
-        for (; _First != _Last; ++_First) {
-            if constexpr (!is_same_v<_Size, size_type>) {
-                if (_My_data._Mysize == _My_data._Myres) { // Need to grow
-                    if (_My_data._Mysize == max_size()) {
-                        _Xlen_string(); // result too long
-                    }
 
-                    _Elem* const _Old_ptr   = _My_data._Myptr();
-                    size_type _New_capacity = _Calculate_growth(_My_data._Mysize + 1);
-                    const pointer _New_ptr  = _Allocate_for_capacity(_Al, _New_capacity); // throws
+        constexpr bool _Can_construct_by_memcpy = _Is_specialization_v<_Traits, char_traits> && _Is_EcharT<_Elem>
+                                               && _Iterator_is_contiguous<_Iter> && !_Iterator_is_volatile<_Iter>
+                                               && is_integral_v<_Iter_value_t<_Iter>>
+                                               && sizeof(_Iter_value_t<_Iter>) == sizeof(_Elem);
+
+        if constexpr (_Can_construct_by_memcpy) {
+            _STL_INTERNAL_STATIC_ASSERT(is_same_v<_Size, size_type>);

```

Comment:
We can't derive this from the preconditions of the function. Indicates lack of test coverage for non-sized non-common contiguous ranges like the classic:

```c++
struct c_string_view : std::ranges::view_interface<c_string_view> {
  struct sentinel {
    friend bool operator==(const char* p, sentinel) { return *p == '\0'; }
  };

  /* implicit */ c_string_view(const char* ptr) : first{ptr} {}

  const char* begin() const { return first; }
  sentinel end() const { return {}; }
private:
  const char* first;
};
```

Reply:
Luckily, I'm wrong about test coverage; there's no code path that calls this function with a contiguous iterator, a different sentinel type, and an invalid `_Count`. "Different sentinel type" could only have come from the `from_range_t` tagged constructor, and that constructor always calculates the length of forward ranges. We could make `_Construct_from_iter` more general - and remove this assert - or we can strengthen the preconditions (they can always be relaxed later).

Reply:
```c++
        // Pre: if _Iter models forward_iterator or meets the Cpp17ForwardIterator requirements,
        //      then is_same_v<_Size, size_type> holds.
```
works.